### PR TITLE
[IE Samples] Fixed hanging of samples if InferImpl() throws exception

### PR DESCRIPTION
**Pull request by mgrdv**
Details:
 - Added rethrow an exception in samples 

Tickets:
 - 81939

### CODE REVIEW

File: samples/cpp/classification_sample_async/main.cpp [hunk 1]
```diff
@@ -170,7 +170,7 @@ int main(int argc, char* argv[]) {
         // -------- Step 10. Do asynchronous inference --------
         infer_request.set_callback([&](std::exception_ptr ex) {
             if (ex)
-                throw ex;
+                std::rethrow_exception(ex);
```
Comment (Wovchena) [inline samples/cpp/classification_sample_async/main.cpp:173]: You need to propagate the exception to the main thread

Comment (vladimir-dudnik) [reply samples/cpp/classification_sample_async/main.cpp:173]: @Wovchena could you suggest what exactly should be done?

Comment (Wovchena) [reply samples/cpp/classification_sample_async/main.cpp:173]: Same try catch as in https://github.com/openvinotoolkit/open_model_zoo/pull/3376

Comment (apankratovantonp) [reply samples/cpp/classification_sample_async/main.cpp:173]: Please use `std::promise<void>/std::future<void>` to wait for execution or propagate exception to main thread. 
Please use `OVInferRequestCallbackTests.canStartSeveralAsyncInsideCompletionCallbackWithSafeDtor` as example

Comment (mgrdv) [reply samples/cpp/classification_sample_async/main.cpp]: @apankratovantonp , please see current variant using variable for checking status of exception. Sample gets error and not hanging without std::future

Comment (mgrdv) [reply samples/cpp/classification_sample_async/main.cpp]: @apankratovantonp or do you want to realize it using std::future instead of current solution?

**Review by vladimir-dudnik — APPROVED** [review]

File: samples/cpp/classification_sample_async/main.cpp (hunk 1 above)
Comment (apankratovantonp) [reply samples/cpp/classification_sample_async/main.cpp]: The last solution is enough to propagate exception to main thread.

**Review by apankratovantonp — APPROVED** [review]

**Pull request merged**